crates/rethnet_evm/src/blockchain/local.rs

```diff
@@ -0,0 +1,219 @@
+        let total_difficulty = genesis_block.header.difficulty;
+        let storage = ContiguousBlockchainStorage::with_block(genesis_block, total_difficulty);
+
+        Self { storage }
+    }
+}
+
+impl Blockchain for LocalBlockchain {
+    type Error = BlockchainError;
+
+    fn block_by_hash(&self, hash: &B256) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        Ok(self.storage.block_by_hash(hash).cloned())
+    }
+
+    fn block_by_number(&self, number: &U256) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        let number = usize::try_from(number).map_err(|_| BlockchainError::BlockNumberTooLarge)?;
+
+        Ok(self.storage.blocks().get(number).cloned())
+    }
+
+    fn block_by_transaction_hash(
+        &self,
+        transaction_hash: &B256,
+    ) -> Result<Option<Arc<DetailedBlock>>, Self::Error> {
+        Ok(self
+            .storage
+            .block_by_transaction_hash(transaction_hash)
+            .cloned())
+    }
+
+    fn last_block(&self) -> Result<Arc<DetailedBlock>, Self::Error> {
+        Ok(self
+            .storage
+            .blocks()
+            .last()
+            .expect("A genesis block is always present")
+            .clone())
+    }
+
+    fn last_block_number(&self) -> U256 {
+        // The block number of the genesis block is 0, so subtract one
+        U256::from(self.storage.blocks().len() - 1)
+    }
+
+    fn total_difficulty_by_hash(&self, hash: &B256) -> Result<Option<U256>, Self::Error> {
+        Ok(self.storage.total_difficulty_by_hash(hash).cloned())
+    }
+}
+
+impl BlockchainMut for LocalBlockchain {
+    type Error = BlockchainError;
+
+    fn insert_block(&mut self, block: DetailedBlock) -> Result<Arc<DetailedBlock>, Self::Error> {
+        let last_block = self.last_block()?;
+
+        let next_block_number = last_block.header.number + U256::from(1);
+        if block.header.number != next_block_number {
+            return Err(BlockchainError::InvalidBlockNumber {
+                actual: block.header.number,
+                expected: next_block_number,
+            });
+        }
+
+        if block.header.parent_hash != last_block.header.hash() {
+            return Err(BlockchainError::InvalidParentHash);
+        }
+
+        let previous_total_difficulty = self
+            .storage
+            .total_difficulties()
+            .last()
+            .expect("Storage always contains at least one block");
+
+        let total_difficulty = previous_total_difficulty + block.header.difficulty;
+
+        // SAFETY: The block number is guaranteed to be unique, so the block hash must be too.
+        let block = unsafe { self.storage.insert_block_unchecked(block, total_difficulty) };
+
+        Ok(block.clone())
+    }
+}
+
+impl BlockHashRef for LocalBlockchain {
+    type Error = BlockchainError;
+
+    fn block_hash(&self, number: U256) -> Result<B256, Self::Error> {
+        // Question: Do we need to support block number larger than usize::MAX
```

Comment:
@fvictorio, @alcuadrado This question was intended for you. Theoretically block numbers can be larger than `u64::MAX`, but the amount of memory that can be addresses on a computer cannot. Hence, my question: will we ever need to support block numbers that are larger?

Reply:
We computed it, and it's safe to assume that they feet in a u64, even if blocks were created every second.